libp2p/utils/future.nim

```diff
@@ -31,3 +32,19 @@ proc anyCompleted*[T](futs: seq[Future[T]]): Future[Future[T]] {.async.} =
 
     let index = requests.find(raceFut)
     requests.del(index)
+
+proc raceAndCancelPending*(
+    futs: seq[SomeFuture]
+): Future[void] {.async: (raises: [ValueError, CancelledError]).} =
+  ## Executes a race between the provided sequence of futures.
+  ## Cancels any remaining futures that have not yet completed.
+  ##
+  ## - `futs`: A sequence of futures to race.
+  ##
+  ## Raises:
+  ## - `ValueError` if the sequence of futures is empty.
+  ## - `CancelledError` if the operation is canceled.
+  try:
+    discard await race(futs)
+  finally:
```

Comment:
there is no need for finally .. in particular, if `race` is cancelled, it should not cancel `futs` to maintain the `race` contract in general - ValueError means there are no futures, so nothing to cancel either